### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by dmuhs**
This PR will
- add Sphinx as a documentation system
- update the docs with autodoc instructions for all components of the code base
- add basic installation instructions to the docs
- fix a minor indentation issue in the `Swarm.listen` docstring

With this PR merged, documentation can be built locally by entering the docs folder and using the Makefile:

```
cd docs && make html
```

The output can be found in the `build` directory, in the case of `html` docs, it can be opened at `docs/build/html/index.html`. A variety of other documentation formats are available and can be checked out by using the Makefile's help command:

```
$ make help
Sphinx v2.2.0
Please use `make target' where target is one of
  html        to make standalone HTML files
  dirhtml     to make HTML files named index.html in directories
  singlehtml  to make a single large HTML file
  pickle      to make pickle files
  json        to make JSON files
  htmlhelp    to make HTML files and an HTML help project
  qthelp      to make HTML files and a qthelp project
  devhelp     to make HTML files and a Devhelp project
  epub        to make an epub
  latex       to make LaTeX files, you can set PAPER=a4 or PAPER=letter
  latexpdf    to make LaTeX and PDF files (default pdflatex)
  latexpdfja  to make LaTeX files and run them through platex/dvipdfmx
  text        to make text files
  man         to make manual pages
  texinfo     to make Texinfo files
  info        to make Texinfo files and run them through makeinfo
  gettext     to make PO message catalogs
  changes     to make an overview of all changed/added/deprecated items
  xml         to make Docutils-native XML files
  pseudoxml   to make pseudoxml-XML files for display purposes
  linkcheck   to check all external links for integrity
  doctest     to run all doctests embedded in the documentation (if enabled)
  coverage    to run coverage check of the documentation (if enabled)
```

New documentation can be added by simply manipulating the code components' docstrings, or explicitly by editing the `rst` files in the docs folder. These files mirror 1:1 the module structure of the library at the moment and should be adapted in case larger refactorings take place. This makes it easy for developers already familiar with the project structure to also contribute to the documentation if further explanation is needed.

The goal of this PR is to lay the foundation for further exhaustive documentation of each code component (ref. #318, #90, #35), as well as improving docstring quality and guiding future refactors leading `py-libp2p` to a state where we can remove the WIP label from it. :slightly_smiling_face:

### CODE REVIEW

**Comment by dmuhs** [comment]
Added fix for mypy error, taken from https://github.com/libp2p/py-libp2p/pull/328/commits/d52b093286c9b4c33b26bb9d4af9ee5a24e52c71

**Comment by pipermerriam** [comment]
cc @cburgdorf who is our resident documentation expert.

**Comment by carver** [comment]
For reference, here's an excerpt of the global makefile we use to build docs in Trinity: https://github.com/ethereum/trinity/blob/24213d8f22a03c370313d3a990eea6a2fe1e76e4/Makefile#L43-L57

I'm happy to take a look at this PR tomorrow.

File: libp2p/stream_muxer/mplex/mplex_stream.py [hunk 1]
```diff
@@ -146,7 +146,7 @@ def is_initiator(self) -> bool:
         :return: number of bytes written
         """
         if self.event_local_closed.is_set():
-            raise MplexStreamClosed(f"cannot write to closed stream: data={data}")
+            raise MplexStreamClosed(f"cannot write to closed stream: data={data!r}")
```
Comment (carver) [inline libp2p/stream_muxer/mplex/mplex_stream.py]: I'd recommend this one to be moved to another PR, it should be a very quick :+1:

File: libp2p/utils.py [hunk 2]
```diff
@@ -77,7 +77,7 @@ def encode_delim(msg: bytes) -> bytes:
         raise ParseError(f"`len(msg_bytes)` should not be 0")
     if msg_bytes[-1:] != b"\n":
         raise ParseError(
-            f'`msg_bytes` is not delimited by b"\\n": `msg_bytes`={msg_bytes}'
+            f'`msg_bytes` is not delimited by b"\\n": `msg_bytes`={msg_bytes!r}'
```
Comment (carver) [inline libp2p/utils.py]: This one can also be added to that separate PR.

File: setup.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
         "isort==4.3.21",
         "flake8>=3.7.7,<4.0.0",
     ],
-    "dev": ["tox>=3.13.2,<4.0.0"],
+    "dev": ["tox>=3.13.2,<4.0.0", "Sphinx>=2.2.0,<3.0.0", "sphinx-rtd-theme"],
```
Comment (carver) [inline setup.py]: It can be nice to move this to its own `doc` extra, like:

https://github.com/ethereum/trinity/blob/0ef04653c17d726463ae8a7f0ac01e0dcd930d54/setup.py#L85-L91

File: libp2p/stream_muxer/mplex/mplex_stream.py (hunk 1 above)
Comment (dmuhs) [reply libp2p/stream_muxer/mplex/mplex_stream.py]: It seems like this is already resolved in `master` with d52b093286c9b4c33b26bb9d4af9ee5a24e52c71

File: libp2p/utils.py (hunk 2 above)
Comment (dmuhs) [reply libp2p/utils.py]: Also fixed in d52b093286c9b4c33b26bb9d4af9ee5a24e52c71 on `master` already

File: setup.py (hunk 3 above)
Comment (dmuhs) [reply setup.py]: Fixed in f7403ef349a03350224a21b19caf1343c54a35ec and added doc dependencies to the dev requirements in 817e9c88cfa61ef05d024ae8c7b2f4aaa7c4ca89

File: docs/source/index.rst [hunk 4]
```diff
@@ -0,0 +1,17 @@
+Welcome to py-libp2p's documentation!
```
Comment (cburgdorf) [inline docs/source/index.rst]: We have a few general [documentation guidelines](https://github.com/ethereum/snake-charmers-tactical-manual/blob/master/documentation.md), and we're trying to make sure that all our repos are as close to them as possible.

That's why I'm taking this opportunity to point this out so that we can get these policies involved early on.

It's far from perfect but the project closest to the ideals of this guide line is probably Trinity so I recommend to take a [look at its docs.](https://trinity-client.readthedocs.io/en/latest/)

In particular, it would be nice if we could roughly follow this navigation structure.

image

### General
**Introduction**: The landing page, giving a quick overview about the project, additional pointers.
**Quickstart**: Pretty much what you already have with all the installation instructions
**Release Notes**: Points to the towncrier generated release notes

### Fundamentals

**API**: Source generated API docs as you already have them included in the PR
**Cookbook**: Collection of small recipes, examples. Doesn't need to exist in the beginning
**Guides**: Tutorial style guides. It's ok to just have a link to the *Quickstart* here to begin with

### Community

**Contributing**: Instructions on how to contribute to the project
**Code of Conduct**: The code of conduct that we use across repos

File: docs/source/installation.rst [hunk 5]
```diff
@@ -0,0 +1,50 @@
+.. highlight:: shell
+
+Installation
+============
+
+Stable release
+--------------
+
+To install :code:`py-libp2p`:, run this command in your terminal:
```
Comment (cburgdorf) [inline docs/source/installation.rst]: There's no right or wrong when it comes to the [*narrative perspective*](https://github.com/ethereum/snake-charmers-tactical-manual/blob/master/documentation.md#narrative-perspective) but we chose to use the *first-person "we" form* which means there is rarely ever a place for "you" in our docs.

E.g. see [Trinity's quickstart](https://trinity-client.readthedocs.io/en/latest/guides/quickstart.html) written in *first person "we" form*

**Comment by pacrob** [comment]
Docs reorged and more in #454
